File: assets-source/blog-image-validation/generate_symbol_search_fixture.py
#!/usr/bin/env python3
"""Generate the owned five-symbol fixture used for release screenshot evidence."""

from pathlib import Path
import sys

from reportlab.lib.pagesizes import letter
from reportlab.pdfgen import canvas


output = Path(sys.argv[1] if len(sys.argv) > 1 else "/tmp/polypdf-symbol-search-fixture.pdf")
output.parent.mkdir(parents=True, exist_ok=True)
page_width, page_height = letter
symbol_size = 40
anchors = [(100, 100), (250, 100), (400, 100), (100, 250), (250, 250)]


def draw_symbol(pdf: canvas.Canvas, x: float, top: float, size: float) -> None:
    stroke = max(2, size / 20)
    bottom = page_height - top - size
    pdf.setLineWidth(stroke)
    pdf.rect(x, bottom, size, size, fill=0, stroke=1)
    pdf.line(x, page_height - top, x + size, bottom)
    block = size / 4
    pdf.rect(x + stroke * 2, page_height - top - stroke * 2 - block, block, block, fill=1, stroke=0)


pdf = canvas.Canvas(str(output), pagesize=letter, pageCompression=0)
pdf.setTitle("Owned PolyPDF Symbol Search Fixture")
pdf.setAuthor("PolyPDF Evidence Lab")
for anchor in anchors:
    draw_symbol(pdf, *anchor, symbol_size)
draw_symbol(pdf, 180, 430, 260)
pdf.save()
print(output)

File: assets-source/blog-image-validation/generate-symbol-search-fixture.mjs
#!/usr/bin/env node
import { createRequire } from 'node:module';
import { writeFile } from 'node:fs/promises';
import { join, resolve } from 'node:path';

const argv = process.argv.slice(2);
const valueAfter = (flag, fallback) => {
  const index = argv.indexOf(flag);
  return index >= 0 ? argv[index + 1] : fallback;
};
const appRoot = resolve(valueAfter('--app-root', '/tmp/polypdf-1.4-capture-tQ8oBz/worktree/polypdf'));
const output = resolve(valueAfter('--output', '/tmp/polypdf-symbol-search-fixture.pdf'));
const { PDFDocument, rgb } = createRequire(join(appRoot, 'package.json'))('pdf-lib');

const PAGE_WIDTH = 612;
const PAGE_HEIGHT = 792;
const SIZE = 40;
const LARGE_SIZE = 260;
const anchors = [[100, 100], [250, 100], [400, 100], [100, 250], [250, 250]];
const pdf = await PDFDocument.create();
const page = pdf.addPage([PAGE_WIDTH, PAGE_HEIGHT]);
const black = rgb(0, 0, 0);
const flip = (appY) => PAGE_HEIGHT - appY;

function drawSymbol(ax, ay, size) {
  const stroke = Math.max(2, size / 20);
  page.drawRectangle({
    x: ax,
    y: flip(ay + size),
    width: size,
    height: size,
    borderColor: black,
    borderWidth: stroke
  });
  page.drawLine({
    start: { x: ax, y: flip(ay) },
    end: { x: ax + size, y: flip(ay + size) },
    thickness: stroke,
    color: black
  });
  const block = size / 4;
  page.drawRectangle({
    x: ax + stroke * 2,
    y: flip(ay + stroke * 2 + block),
    width: block,
    height: block,
    color: black
  });
}

for (const [x, y] of anchors) drawSymbol(x, y, SIZE);
drawSymbol(180, 430, LARGE_SIZE);
await writeFile(output, await pdf.save());
process.stdout.write(`${output}\n`);
